File: frontend/src/utils/passwordValidator.ts
export const FORBIDDEN_WORDS = [
    "parking", "parkingplus",
    "student", "uczelnia", "projekt", "admin", "polska", "system"
];
export const COMMON_PATTERNS = [
    "123456", "qwerty", "password", "abcdef", "123456789", "haslo123", "admin123", "polska123", "qwertyuiop"
];

export interface PasswordValidationResult {
    isValid: boolean;
    errors: string[];
}

async function sha1(message: string): Promise<string> {
    const msgBuffer = new TextEncoder().encode(message);
    const hashBuffer = await crypto.subtle.digest('SHA-1', msgBuffer);
    const hashArray = Array.from(new Uint8Array(hashBuffer));
    return hashArray.map(b => b.toString(16).padStart(2, '0')).join('').toUpperCase();
}

export async function checkHIBP(password: string): Promise<boolean> {
    try {
        const hash = await sha1(password);
        const prefix = hash.substring(0, 5);
        const suffix = hash.substring(5);

        const response = await fetch(`https://api.pwnedpasswords.com/range/${prefix}`);
        if (!response.ok) return false;

        const text = await response.text();
        const lines = text.split('\n');

        return lines.some(line => line.split(':')[0] === suffix);
    } catch (e) {
        console.error("HIBP check failed", e);
        return false;
    }
}

export interface PasswordValidationRules {
    minLength: boolean;
    noUserData: boolean;
    noForbiddenWords: boolean;
    noCommonPatterns: boolean;
}

export function checkPasswordRules(
    password: string,
    userData: string[] = []
): PasswordValidationRules {
    const lowerPassword = password.toLowerCase();

    return {
        minLength: password.length >= 12,
        noUserData: !userData.some(data =>
            data && data.trim() && lowerPassword.includes(data.toLowerCase().trim())
        ),
        noForbiddenWords: !FORBIDDEN_WORDS.some(word => lowerPassword.includes(word)),
        noCommonPatterns: !COMMON_PATTERNS.some(pattern => lowerPassword.includes(pattern))
    };
}

export async function validatePassword(
    password: string,
    userData: string[] = []
): Promise<PasswordValidationResult> {
    const errors: string[] = [];
    const rules = checkPasswordRules(password, userData);

    if (!rules.minLength) errors.push('logins.errors.auth.passwordTooShort12');
    if (!rules.noUserData) errors.push('logins.errors.auth.passwordContainsUserData');
    if (!rules.noForbiddenWords) errors.push('logins.errors.auth.passwordContainsForbiddenWord');
    if (!rules.noCommonPatterns) errors.push('logins.errors.auth.passwordTooCommon');

    // Only check HIBP if other rules pass to save API calls
    if (errors.length === 0) {
        const isPwned = await checkHIBP(password);
        if (isPwned) {
            errors.push('logins.errors.auth.passwordPwned');
        }
    }

    return {
        isValid: errors.length === 0,
        errors
    };
}

File: backend/src/main/kotlin/com/parkingplus/security/validation/PasswordValidator.kt
package com.parkingplus.security.validation

import org.springframework.stereotype.Component
import org.springframework.web.client.RestTemplate
import java.security.MessageDigest
import java.util.*

@Component
class PasswordValidator {

    private val restTemplate = RestTemplate()
    private val forbiddenWords = listOf(
        "parking", "parkingplus", "politechnika",
        "student", "uczelnia", "projekt", "admin", "polska", "system"
    )

    fun validate(password: String, userData: List<String>) {
        // 1. Minimum length
        if (password.length < 12) {
            throw IllegalArgumentException("Hasło musi mieć co najmniej 12 znaków")
        }

        // 2. Context-aware checks
        val normalizedPassword = password.lowercase(Locale.getDefault())

        // User data (name, surname, email)
        userData.forEach { data ->
            if (data.isNotBlank() && normalizedPassword.contains(data.lowercase(Locale.getDefault()))) {
                throw IllegalArgumentException("Hasło nie może zawierać Twoich danych osobowych")
            }
        }

        // App context words
        forbiddenWords.forEach { word ->
            if (normalizedPassword.contains(word)) {
                throw IllegalArgumentException("Hasło zawiera niedozwolone słowa związane z aplikacją")
            }
        }

        // 3. Common patterns (simplified)
        val commonPatterns = listOf(
            "123456", "qwerty", "password", "abcdef", "123456789", "zaq12wsx", 
            "marcin", "adam123", "haslo123", "admin123", "polska123", "qwertyuiop"
        )
        commonPatterns.forEach { pattern ->
            if (normalizedPassword.contains(pattern)) {
                throw IllegalArgumentException("Hasło zawiera zbyt proste i popularne sekwencje znaków")
            }
        }

        // 4. HIBP Leak Check
        if (isPasswordPwned(password)) {
            throw IllegalArgumentException("To hasło pojawiło się w znanych wyciekach danych. Wybierz inne.")
        }
    }

    private fun isPasswordPwned(password: String): Boolean {
        val sha1 = sha1(password)
        val prefix = sha1.substring(0, 5)
        val suffix = sha1.substring(5).uppercase(Locale.getDefault())

        val url = "https://api.pwnedpasswords.com/range/$prefix"
        val response = restTemplate.getForObject(url, String::class.java) ?: return false

        return response.lineSequence().any { line ->
            line.split(":")[0] == suffix
        }
    }

    private fun sha1(input: String): String {
        val md = MessageDigest.getInstance("SHA-1")
        val bytes = md.digest(input.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }.uppercase(Locale.getDefault())
    }
}
